### PR TITLE
UML 1080: Remove Lpa API

### PR DESCRIPTION
**Pull request by SeemaMenon**
# Purpose

https://opgtransform.atlassian.net/browse/UML-1080

Fixes UML-1080

## Approach

_Explain how your code addresses the purpose of the change_

## Learning

_Any tips and tricks, blog posts or tools which helped you. Plus anything notable you've discovered about the Use a Lasting Power of Attorney service_

## Checklist

* [X] I have performed a self-review of my own code
* [X] I have added relevant logging with appropriate levels to my code
* [ ] I have updated documentation (Confluence/GitHub wiki/tech debt doc) where relevant
* [X] I have added tests to prove my work
* [X] I have added welsh translation tags and updated translation files
* [ ] The product team have tested these changes

### CODE REVIEW

File: service-api/app/features/context/Acceptance/AccountContext.php [hunk 1]
```diff
@@ -1305,4 +1310,32 @@ public function iShouldBeToldThatABadRequestWasMade()
     {
         $this->ui->assertSession()->statusCodeEquals(StatusCodeInterface::STATUS_INTERNAL_SERVER_ERROR);
     }
+
+    /**
+     * @When /^I confirm removal of the LPA$/
+     */
+    public function iConfirmRemovalOfTheLPA()
```
Comment (GemTay) [inline service-api/app/features/context/Acceptance/AccountContext.php:1317]: All tests to do with deleting an LPA should live in the LpaContext class rather than the AccountContext

File: service-api/app/features/context/Integration/LpaContext.php [hunk 2]
```diff
@@ -1294,4 +1295,98 @@ public function iCanSeeTheNameOfTheOrganisationThatViewedTheLPA()
     {
         // Not needed for this context
     }
+
+    /**
+     * @When /^I confirm removal of the LPA/
+     */
+    public function iConfirmRremovalOfTheLpa()
```
Comment (GemTay) [inline service-api/app/features/context/Integration/LpaContext.php:1302]: small typo in the word 'Removal'

File: service-api/app/src/App/src/Service/Lpa/LpaService.php [hunk 3]
```diff
@@ -330,4 +330,47 @@ private function attorneyStatus(array $attorney): int
 
         return self::ACTIVE_ATTORNEY;
     }
+
+    /**
+     * @param string $actorLpaToken
+     * @return array
+     */
+    public function removeLPaFromUserLpaActorMap(string $actorLpaToken): array
```
Comment (GemTay) [inline service-api/app/src/App/src/Service/Lpa/LpaService.php:338]: typo 'removeLPA...'

File: service-api/app/src/App/src/Service/Lpa/LpaService.php [hunk 4]
```diff
@@ -330,4 +330,47 @@ private function attorneyStatus(array $attorney): int
...
+    {
+        $userActorLpa = $this->userLpaActorMapRepository->get($actorLpaToken);
```
Comment (GemTay) [inline service-api/app/src/App/src/Service/Lpa/LpaService.php:340]: Would it be worth also checking that the logged in user Id matches the UserId in the UserLpaActorMap for the LPA that they are trying to delete? This would increase the security of the feature

File: service-api/app/test/AppTest/DataAccess/DynamoDb/UserLpaActorMapTest.php [hunk 5]
```diff
@@ -208,22 +208,37 @@ public function can_delete_map()
         $testToken = 'test-token';
 
         $this->dynamoDbClientProphecy->deleteItem(Argument::that(function(array $data) use ($testToken) {
-            $this->assertArrayHasKey('TableName', $data);
-            $this->assertEquals(self::TABLE_NAME, $data['TableName']);
-
-            //---
+            //$this->assertArrayHasKey('TableName', $data);
```
Comment (GemTay) [inline service-api/app/test/AppTest/DataAccess/DynamoDb/UserLpaActorMapTest.php:211]: comment to be removed

File: service-api/app/test/AppTest/Service/Lpa/LpaServiceTest.php [hunk 6]
```diff
@@ -68,6 +69,19 @@ private function getLpaService(): LpaService
         );
     }
 
+    private function getViewerCodeService(): ViewerCodeService
+    {
+        $viewerCodeRepoProphecy = $this->prophesize(ViewerCodesInterface::class);
+        $userActorLpaRepoProphecy = $this->prophesize(UserLpaActorMapInterface::class);
+        $lpaServiceProphecy = $this->prophesize(LpaService::class);
+
+        $service = new ViewerCodeService(
```
Comment (GemTay) [inline service-api/app/test/AppTest/Service/Lpa/LpaServiceTest.php:78]: Missing return statement. This can be simplified by changing $service = to "return"

File: service-front/app/src/Actor/templates/actor/check-access-codes.html.twig [hunk 7]
```diff
@@ -73,7 +73,9 @@
                                     {% set id_counter = ( id_counter | default(0) ) + 1 %}
                                     {% if not check_if_code_has_expired(code.Expires) %}
                                         {% if not check_if_code_is_cancelled(code) %}
-                                            {{ include('@actor/partials/check-code-details.html.twig', {'id_counter': id_counter }) }}
+                                            {% if not check_if_user_lpa_actor_set(code.UserLpaActor) %}
```
Comment (GemTay) [inline service-front/app/src/Actor/templates/actor/check-access-codes.html.twig:76]: Does this mean that co-actors on the same LPA would not longer be able to see the codes made by the actor who has deleted the LPA? Was this the intended behaviour in the ticket?

File: service-front/app/src/Common/src/View/Twig/LpaExtension.php [hunk 8]
```diff
@@ -247,4 +248,19 @@ private function getDateFormatter(string $locale, ?string $pattern): IntlDateFor
...
+     * Checks whether the user lpa actor token is removed
+     *
+     * @param string|null $token
+     * @return bool|null
+     * @throws Exception
+     */
+    public function isUserLpaActorSet(string $token): ?bool
+    {
+        if (empty($token)) {
+            return true;
```
Comment (GemTay) [inline service-front/app/src/Common/src/View/Twig/LpaExtension.php:262]: The logic of the return statements here is a little confusing. To match the function name, if the token is empty it should return false to read "isUserLpaActorSet" -> (no) return false. If the token is set, you should return true rather than null

File: service-front/app/test/CommonTest/View/Twig/LpaExtensionTest.php [hunk 9]
```diff
@@ -465,4 +466,42 @@ public function it_returns_donor_name_from_donor_nameDob_string()
...
+            'Organisation' => 'bank',
+            'UserLpaActor' => 'be89ba48-0415-4ec7-9ca1-f5e8f7f8b224',
+            'ViewerCode' => 'KGC6CB6GNXHC',
+            'Viewed' => false,
+        ];
+
+        $extension = new LpaExtension();
+        $isSet = $extension->isUserLpaActorSet($code['UserLpaActor']);
+
+        $this->assertEquals(false, $isSet);
```
Comment (GemTay) [inline service-front/app/test/CommonTest/View/Twig/LpaExtensionTest.php:505]: add another test to test both scenarios, when the token is empty and when its not

**Review by GemTay — CHANGES_REQUESTED** [review]
Few changes to be made

File: service-front/app/test/CommonTest/View/Twig/LpaExtensionTest.php (hunk 9 above)
Comment (SeemaMenon) [reply service-front/app/test/CommonTest/View/Twig/LpaExtensionTest.php:505]: Is included  
it_checks_if_user_LPA_actor_is_not_set()
it_checks_if_user_LPA_actor_is_set()

[am unclear the suggestion here]

File: service-front/app/src/Actor/templates/actor/check-access-codes.html.twig (hunk 7 above)
Comment (SeemaMenon) [reply service-front/app/src/Actor/templates/actor/check-access-codes.html.twig]: Yes is my understanding from the kickoff conversation we had a week back [Shall get this checked Sarah again].

Removal of an LPA from account by user means removing the associated viewer codes too. But we agreed on the call the viewer code record will not be removed, but will be disassociated from the actor, so that they/or any actor do not see it on the check access code page. This area is not user tested yet! Also the chances of a user removing an LPA [where access code has been generated] is less likely is the assumption, hence requirements in this area was not explicitly mentioned and still fluid until tested I guess.

But if a viewer code [of a removed LPA] was already shared with an organisation, they would be able to view the LPA if it's not expired or cancelled.

File: service-api/app/src/App/src/Service/Lpa/LpaService.php [hunk 10]
```diff
@@ -330,4 +331,53 @@ private function attorneyStatus(array $attorney): int
...
+            $this->logger->notice(
+                'User actor lpa record  not found for actor token {Id}',
+                ['Id' => $actorLpaToken]
+            );
+            throw new NotFoundException('User actor lpa record  not found for actor token - ' . $actorLpaToken);
+        }
+
+        // Ensure the passed userId matches the passed token
+        if ($userId !== $userActorLpa['UserId']) {
+            return null;
```
Comment (GemTay) [inline service-api/app/src/App/src/Service/Lpa/LpaService.php:355]: Probably more suitable to throw an exception here, since its suspicious activity if a user has manages to delete an LPA from an account which does not have that LPA added

File: service-api/app/test/AppTest/Service/Lpa/LpaServiceTest.php [hunk 11]
```diff
@@ -830,4 +845,126 @@ public function can_not_find_actor_who_is_an_inactive_attorney()
...
+    public function remove_lpa_from_user_lpa_actor_map_when_user_id_does_not_match()
+    {
+        $userActorLpa = [
+            'SiriusUid' => '700000055554',
+            'Added' => (new DateTime('2020-01-01'))->format('Y-m-d\TH:i:s.u\Z'),
+            'Id' => '2345Token0123',
+            'ActorId' => '1',
+            'UserId' => '6789',
+        ];
+        $viewerCodes = null;
```
Comment (GemTay) [inline service-api/app/test/AppTest/Service/Lpa/LpaServiceTest.php:958]: doesn't look like this variable is used

**Review by GemTay — CHANGES_REQUESTED** [review]
just a few more small changes

**Comment by SeemaMenon** [comment]
> service-api/app/test/AppTest/Service/Lpa/LpaServiceTest.php

removed

File: service-api/app/test/AppTest/Service/Lpa/LpaServiceTest.php (hunk 11 above)
Comment (SeemaMenon) [reply service-api/app/test/AppTest/Service/Lpa/LpaServiceTest.php:958]: removed

File: service-api/app/src/App/src/Service/Lpa/LpaService.php (hunk 10 above)
Comment (SeemaMenon) [reply service-api/app/src/App/src/Service/Lpa/LpaService.php:355]: Done :)

File: service-api/app/src/App/src/Service/Lpa/LpaService.php [hunk 12]
```diff
@@ -330,4 +331,57 @@ private function attorneyStatus(array $attorney): int
...
+                ['Id' => $actorLpaToken]
+            );
+            throw new NotFoundException('User actor lpa record  not found for actor token - ' . $actorLpaToken);
+        }
+
+        // Ensure the passed userId matches the passed token
+        if ($userId !== $userActorLpa['UserId']) {
+            $this->logger->notice(
+                'User Id passed does not match the user in userActorLpaMap for token - ' . $actorLpaToken,
+                ['Id' => $actorLpaToken]
```
Comment (GemTay) [inline service-api/app/src/App/src/Service/Lpa/LpaService.php:357]: This log message doesn't look the correct format as there is no {Id} wrapped in curly braces in the message. Think it should be 
`$this->logger->notice(
                'User Id passed does not match the user in userActorLpaMap for token {Id}',
                ['Id' => $actorLpaToken]
            );`

**Review by GemTay — CHANGES_REQUESTED** [review]
last one, sorry!

File: service-api/app/src/App/src/Service/Lpa/LpaService.php (hunk 12 above)
Comment (SeemaMenon) [reply service-api/app/src/App/src/Service/Lpa/LpaService.php:357]: :( :( :( MY BAD !

**Comment by SeemaMenon** [comment]
> last one, sorry!

@GemTay - Really sorry for this! Corrected ....

**Review by GemTay — APPROVED** [review]
Looks good to me :)

File: service-front/app/src/Actor/templates/actor/check-access-codes.html.twig [hunk 13]
```diff
@@ -77,7 +77,9 @@
                                     {% set id_counter = ( id_counter | default(0) ) + 1 %}
                                     {% if not check_if_code_has_expired(code.Expires) %}
                                         {% if not check_if_code_is_cancelled(code) %}
-                                            {{ include('@actor/partials/check-code-details.html.twig', {'id_counter': id_counter }) }}
+                                            {% if check_if_user_lpa_actor_set(code.UserLpaActor) %}
```
Comment (cooperaj) [inline service-front/app/src/Actor/templates/actor/check-access-codes.html.twig:80]: 3 nested if's? 

Probs a good idea to rationalise these down to 1

File: service-front/app/src/Common/src/View/Twig/LpaExtension.php [hunk 14]
```diff
@@ -247,4 +248,19 @@ private function getDateFormatter(string $locale, ?string $pattern): IntlDateFor
...
+     * @param string|null $token
+     * @return bool|null
+     * @throws Exception
+     */
+    public function isUserLpaActorSet(string $token): ?bool
+    {
+        if (!empty($token)) {
+            return true;
+        }
+        return false;
```
Comment (cooperaj) [inline service-front/app/src/Common/src/View/Twig/LpaExtension.php:264]:
```suggestion
        return !empty($token);
```

Comment (cooperaj) [reply service-front/app/src/Common/src/View/Twig/LpaExtension.php:264]: But given the simplicity of this TwigFunction I question the need for it. Can this check not be done in the twig file using built-in twig tests?

Comment (SeemaMenon) [reply service-front/app/src/Common/src/View/Twig/LpaExtension.php]: Removed the function in LPAExtensions for empty check and included in twig.

File: service-front/app/src/Actor/templates/actor/check-access-codes.html.twig (hunk 13 above)
Comment (SeemaMenon) [reply service-front/app/src/Actor/templates/actor/check-access-codes.html.twig]: still kept the functions to check if_code_expired and if_code_cancelled, and combined the checks in one loop in the twig.

**Review by cooperaj — APPROVED** [review]
